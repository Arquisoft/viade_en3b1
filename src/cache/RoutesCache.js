import { loadAllRoutes } from '../handler/RouteHandler';
import { loadMedia } from '../handler/MediaHandler';

export default {
    first: true,
    routes: [],
    addRouteToCache(route) {
        if (route && !this.routes.find((obj) => route.name === obj.name)) {
            this.routes.push(route);
        }
    },
    async getRoutes() {
        if (this.routes.length === 0 || this.first) {
            this.routes = await loadAllRoutes();
            this.first = false;
        }
        return this.routes;
    },
    async getRouteById(id) {
        if(this.routes.length === 0) {
            await this.getRoutes();
        }
        let selectedroute = this.routes.find((r) => r.getId() === id);
        if (selectedroute) {
            await selectedroute.calculateElevation();
            await new Promise((r) => setTimeout(r, 1000));
            return selectedroute;
        } else {
            return -1;
        }
        
        
    },
    clear() {
        this.routes = [];
        this.first = true;
    }
};