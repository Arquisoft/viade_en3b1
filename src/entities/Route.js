import { v4 as uuid } from "uuid";
import RouteElement from "./RouteElement";

class Route {

    /**
     * Constructor for a new Route. 
     * An id will be automatically generated (uuid v4).
     * TotalDistance will also be automatically generated.
     * 
     * RouteElements have name, latitude, longitude and elevation.
     * 
     * @param {String} name Name of the route.
     * @param {Date} date Date of the route.
     * @param {String} description Description of the route.
     * @param {Array} routeElements List of RouteElement objects, which are the trackpoints.
     * @param {Array} comments List of comments
     * @param {Array<String>} media List of urls of media (on the POD)
     */
    constructor(name, description, routeElements, comments, media, date) {
        this.id = uuid().toString();
        this.name = name;
        this.date = date;
        this.description = description;

        this.media = media;
        this.totalDistance = null;

        if (comments === null) {
            this.comments = [];
        } else {
            this.comments = comments;
        }

        if (routeElements[0] instanceof RouteElement) {
            this.routeElements = routeElements;
        } else {
            this.routeElements = generateRouteElements(routeElements);
        }
    }

    getId() {
        return this.id;
    }

    getName() {
        return this.name;
    }

    getDate() {
        return this.date;
    }

    getTime() {
        return this.time;
    }

    getDescription() {
        return this.description;
    }

    getRouteElements() {
        return this.routeElements;
    }

    getTotalDistance() {
        var total = 0;
        var aux = 0;

        for (let i = 1; i < this.routeElements.length; i++) {
            aux = this.routeElements[i].getLatitude() - this.routeElements[i - 1].getLatitude();
            total += this.addDistance(aux);
            aux = this.routeElements[i].getLongitude() - this.routeElements[i - 1].getLongitude();
            total += this.addDistance(aux);
            aux = this.routeElements[i].getElevation() - this.routeElements[i - 1].getElevation();
            total += this.addDistance(aux);
        }

        return total;
    }

    addDistance(aux) {
        if (aux < 0) {
            return -aux;
        } else {
            return aux;
        }
    }

    getComments() {
        return this.comments;
    }

    getMedia() {
        return this.media;
    }

    addMedia(media) {
        this.media.push(media);
    }

    addRouteElement(routeElement) {
        this.routeElements.push(routeElement);
    }

    getJsonLD() {
        let routePointsJson = [];
        this.routeElements.forEach((p) => routePointsJson.push(p.toJsonLatLng()));
        return JSON.stringify(
            {
                "@context": {
                    "@version": 1.1,
                    "comments": {
                        "@container": "@list",
                        "@id": "viade:comments"
                    },
                    "description": {
                        "@id": "schema:description",
                        "@type": "xs:string"
                    },
                    "media": {
                        "@container": "@list",
                        "@id": "viade:media"
                    },
                    "name": {
                        "@id": "schema:name",
                        "@type": "xs:string"
                    },
                    "points": {
                        "@container": "@list",
                        "@id": "viade:points"
                    },
                    "latitude": {
                        "@id": "schema:latitude",
                        "@type": "xs:double"
                    },
                    "longitude": {
                        "@id": "schema:longitude",
                        "@type": "xs:double"
                    },
                    "elevation": {
                        "@id": "schema:elevation",
                        "@type": "xsd:double"
                    },
                    "author": {
                        "@id": "schema:author",
                        "@type": "@id"
                    },
                    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
                    "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
                    "schema": "http://schema.org/",
                    "viade": "http://arquisoft.github.io/viadeSpec/",
                    "xsd": "http://www.w3.org/2001/XMLSchema#"
                },
                "name": this.name,
                "author": this.author,
                "description": this.description,
                "comments": this.comments,
                "media": this.media,
                "points": routePointsJson
            }
        );
    }

}

/**
 * Method that receives a list of points and returns
 * a list of RouteElements
 * @param {Array} points List of points containing name, latitude, longitude, elevation. 
 */
function generateRouteElements(points) {
    let routeElements = [];
    points.forEach((p) => routeElements.push(new RouteElement(p.lat, p.lng)));
    return routeElements;
}

export default Route;