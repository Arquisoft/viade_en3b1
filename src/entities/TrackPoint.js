class TrackPoint {

    /**
     * 
     * @param {Number} latitude 
     * @param {Number} longitude 
     * @param {String} name 
     * @param {Number} elevation 
     */
    constructor(latitude, longitude, name, elevation) {

        this.latitude = latitude;
        this.longitude = longitude;

        if (!name) {
            this.name = "";
        } else {
            this.name = name;
        }

        if(!elevation){
            this.elevation = "";
        } else {
            this.elevation = elevation;
        }
    }

    getName() {
        return this.name;
    }

    getLatitude() {
        return this.latitude;
    }

    getElevation() {
        return this.elevation;
    }

    getLongitude() {
        return this.longitude;
    }

    setElevation(num) {
        this.elevation = num;
    }

    toJson() {
        return {
            "latitude": this.latitude,
            "longitude": this.longitude,
            "elevation": this.elevation,
        };
    }
}

export default TrackPoint;