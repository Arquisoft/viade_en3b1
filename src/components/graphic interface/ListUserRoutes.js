import React, { Component } from 'react';
import Grid from '@material-ui/core/Grid';
import MyRouteCard from './MyRouteCard';
import { LoggedIn } from '@solid/react';
import cache from '../../cache/RoutesChache';

export class ListUserRoutes extends Component {
  constructor() {
    super();
    this.state = {
      routes: cache.getRoutesFromCache(),
    };
  }

  sizeFunction() {
    let size = [];
    for (let index = 0; index < 5; index++) {
      size.push(1);
    }
  }

  render() {
    const { routes } = this.state;

    return (
      <Grid container style={{ flexGrow: 1 }} spacing={2}>
      <Grid item xs={12}>
        <Grid container justify="center" spacing={2}>
        { routes.map(each => (
          <Grid item>
          <LoggedIn>
            <MyRouteCard route={each} />
          </LoggedIn>
        </Grid>
        ))}
        </Grid>
      </Grid>
    </Grid>
    )
  }
}

export default ListUserRoutes;