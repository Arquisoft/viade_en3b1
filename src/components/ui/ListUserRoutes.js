import React, { Component } from 'react';
import Grid from '@material-ui/core/Grid';
import MyRouteCard from './MyRouteCard';
import cache from '../../cache/RoutesCache';
import Backdrop from '@material-ui/core/Backdrop';
import CircularProgress from '@material-ui/core/CircularProgress';

export class ListUserRoutes extends Component {
  constructor() {
    super();
    this.state = {
      loading: true,
      routes: [],
    };
  }

  componentDidMount() {
    cache.getRoutes().then((routesList) => {
      this.setState({
        routes: routesList,
        loading: false,
      });
      this.props.toggleBackground();
    });
  }
  
  render() {
    const { routes, loading } = this.state;

    return (
      <Grid container >
        <Grid item xs={'auto'} >
          <Grid container justify="center" spacing={2}>

            <Backdrop style={{color: '#5c5585'}} open={loading} invisible>
              <CircularProgress color="inherit" hidden={loading} />
            </Backdrop>

            {routes.map(each => (
              <Grid item>
                <MyRouteCard route={each} />
              </Grid>
            ))}
          </Grid>
        </Grid>
      </Grid>
    )
  }
}

export default ListUserRoutes;