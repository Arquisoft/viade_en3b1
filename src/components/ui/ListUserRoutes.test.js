import ListUserRoutes from "./ListUserRoutes";
import RoutesCache from "../../cache/RoutesCache";
import Route from '../../entities/Route';
import TrackPoint from '../../entities/TrackPoint';

test('Define ListUserRoutes', () => {
    expect(ListUserRoutes).toBeDefined();
});

test('Renders correctly', () => {
    const wrapper = render(<ListUserRoutes />);
    expect(wrapper).toMatchSnapshot();
});

test('Mounts correctly', () => {
    var route1 = new Route("Route1", "Description 1", [new TrackPoint(1,1), new TrackPoint(2,1)], null, null, null);
    RoutesCache.addRouteToCache(route1);

    const wrapper = mount(<ListUserRoutes />);
});