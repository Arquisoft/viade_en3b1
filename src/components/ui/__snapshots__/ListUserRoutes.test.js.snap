// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div
  aria-hidden="true"
  class="MuiBackdrop-root MuiBackdrop-invisible"
  style="opacity:0;color:#5c5585"
>
  <div
    class="MuiCircularProgress-root MuiCircularProgress-indeterminate"
    hidden=""
    role="progressbar"
    style="width:40px;height:40px"
  >
    <svg
      class="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</div>
`;
